### ISSUE TITLE
mlflow UI not working in iframe[BUG]--cross origin issue

### ISSUE THREAD

**Issue by vasudeva02** [mlflow/mlflow#3583]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **MLflow installed from (source or binary)**:
- **MLflow version (run ``mlflow --version``)**:
- **Python version**:
- **npm version, if running the dev UI**:
- **Exact command to reproduce**:

### Describe the problem
Describe the problem clearly here. Include descriptions of the expected behavior and the actual behavior.

### Code to reproduce issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by mshetty27** [comment]
There is a method in "/src/common/utils/Utils.js" called updatePageTitle. It does a window.parent.postMessage as it should do when communicating with a cross orgin window. However, it accesses "window.parent.location.origin" while setting the "targetOrigin" method parameter. which is the problem. You could put "*" instead but it is not recommended for security reasons. Better would be to specify the exact domain where you are going to embed the mlflow UI in an iframe.

**Comment by vasudeva02** [comment]
Thanks @mshetty27 , it worked !

**Comment by browser-bug** [comment]
As an update, changing `window.parent.location.origin` to `window.location.origin` seems to do the trick. At least in my use case scenario. I don't know if this could be a disruptive action or not but for the little testing I did it seems just fine.

**Comment by Euraxluo** [comment]
Is it fixed?
I also have this problem in the 1.27.0 version

**Comment by daniel-sali** [comment]
I had a similar issue with MLFlow 2.0.1, had to remove the `window.top?.location.href` references from the JS front-end code.

**Comment by ewaldbervoets** [comment]
Does anyone have a solution for MlFlow v2.1.1?

**Comment by ewaldbervoets** [comment]
> I had a similar issue with MLFlow 2.0.1, had to remove the `window.top?.location.href` references from the JS front-end code.

@daniel-sali I'm struggling to get the working in an Iframe, did you remove the references from the bundle or from the react components and rebuild the frontend yourself?

**Comment by daniel-sali** [comment]
@ewaldbervoets Yes, I removed the references from the react components, rebuilt the front-end with `craco` and copied the built files to `mlflow/server/js/build`.

**Comment by ewaldbervoets** [comment]
I was overcomplicating it by trying to change the bundled main.js file. Thanks a lot @daniel-sali , that works!

**Comment by dshakey** [comment]
@ewaldbervoets  @daniel-sali novice question, how do you rebuild from frontend with craco?

**Comment by ewaldbervoets** [comment]
@dshakey I changed the line you linked to `link={window.location.href}`

The frontend is bootstrapped with [create-react-app](https://create-react-app.dev/) and can be rebuild by:
- first installing the dependencies with `yarn install`
- running `yarn build` in the `mlflow/mlflow/server/js` folder. This command uses craco to rebuild the frontend. 

This should generate a new bundle in `/build/static/js` with the changes you made

**Comment by dshakey** [comment]
> @dshakey I changed the line you linked to `link={window.location.href}`
> 
> The frontend is bootstrapped with [create-react-app](https://create-react-app.dev/) and can be rebuild by:
> 
> * first installing the dependencies with `yarn install`
> * running `yarn build` in the `mlflow/mlflow/server/js` folder. This command uses craco to rebuild the frontend.
> 
> This should generate a new bundle in `/build/static/js` with the changes you made

thanks, should have read the Dockerfile first, will give the above a try, just need to sort out error i get with yarn install, get the below and then fails 

```
➤ YN0041: │ websocket-driver@npm:0.6.5: Invalid authentication (as an anonymous user)
➤ YN0041: │ whatwg-fetch@npm:3.6.2: Invalid authentication (as an anonymous user)
➤ YN0041: │ which-module@npm:2.0.0: Invalid authentication (as an anonymous user)
```

**Comment by dshakey** [comment]
thanks all sorted, been awhile since used yarn cli and some of parameters have changed, we use an internal aritfactory so  needed to configure to that

```
    npm install --global yarn && \
    # set yarn config in build folder
    cd mlflow/server/js && \
    yarn config set npmAuthIdent ${ARTYUSER} && \
    yarn config set npmAuthToken ${ARTYTOKEN} && \
    yarn config set npmRegistryServer https://xxxxxxxxx/artifactory/api/npm/npm-virtual/ 


    # fix for working mlflow in iframe
RUN sed 's/window.top?.location.href || window.location.href/window.location.href/g' -i /app/mlflow-${ML_VERSION}/mlflow/server/js/src/experiment-tracking/components/experiment-page/components/header/ExperimentViewHeaderShareButton.tsx && \
    cd mlflow/server/js && \
    yarn install && \
    yarn build && \
```

**Comment by Zhangjiawei-Beinloft** [comment]
when i running `yarn build` in the `mlflow/mlflow/server/js` folder , it report an error like: 
```
craco:  *** Cannot find ESLint plugin (ESLintWebpackPlugin). ***
C:\Users\zhangjiawei16\Desktop\MLflow\mlflow-branch-2.1\mlflow\server\js\craco.config.js:87
    throw new Error('Failed to fix CSS paths!');
```
Has anyone encountered this problem and how can I solve it?
